sync-core/src/main/java/com/cloudant/sync/datastore/encryption/CachingKeyProvider.java

```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright (c) 2015 IBM Cloudant, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package com.cloudant.sync.datastore.encryption;
+
+import java.util.logging.Logger;
+
+/**
+ * This class implements the interface {@link KeyProvider} and it can be used to create an
+ * encrypted datastore.
+ *
+ * Given an user-provided {@link KeyProvider}, it provides an in-memory cache for retrieving an
+ * {@link EncryptionKey}. This improves performance when multiple encrypted datastores are used.
```

Comment:
I think this paragraph should be the first, introductory paragraph as it's the raison d'etre for the class.